### PR TITLE
[E2-2104] Refactoring IAMs

### PR DESCRIPTION
**Pull request by milos1290**
What              | Where/Who
------------------|----------------------------------------
JIRA Issue        | [E2-2104](https://leanplum.atlassian.net/browse/E2-2104)
People Involved   | @milos1290

### CODE REVIEW

File: Leanplum-SDK/Classes/LPMessageTemplates.m [hunk 1]
```diff
@@ -94,31 +84,29 @@ - (id)init
...
-    [[[LPInterstitialMessageTemplate alloc] init] defineActionWithContexts:_contexts];
-    [[[LPWebInterstitialMessageTemplate alloc] init] defineActionWithContexts:_contexts];
-    [[[LPOpenUrlMessageTemplate alloc] init] defineActionWithContexts:_contexts];
-    [[[LPPushAsktoAskMessageTemplate alloc] init] defineActionWithContexts:_contexts];
-    [[[LPRegisterForPushMessageTemplate alloc] init] defineActionWithContexts:_contexts];
-    [[[LPCenterPopupMessageTemplate alloc] init] defineActionWithContexts:_contexts];
-    [[[LPHtmlMessageTemplate alloc] init] defineActionWithContexts:_contexts];
-    [[[LPAppRatingMessageTemplate alloc] init] defineActionWithContexts:_contexts];
-    [[[LPIconChangeMessageTemplate alloc] init] defineActionWithContexts:_contexts];
+    [LPCenterPopupMessageTemplate defineAction];
```
Comment (e7mac) [inline Leanplum-SDK/Classes/LPMessageTemplates.m:87]: I moved away from the pattern of static previously and do not think we should be going back to that

File: Leanplum-SDK/Classes/LPMessageTemplates.m [hunk 2]
```diff
@@ -94,31 +84,29 @@ - (id)init
...
+    [LPAlertMessageTemplate defineAction];
 }
 
 // If notification were enabled by Leanplum's "Push Ask to Ask" or "Register For Push",
 // refreshPushPermissions will do the same registration for subsequent app sessions.
 // refreshPushPermissions is called by [Leanplum start].
 - (void)refreshPushPermissions
 {
-    [[[LPPushAsktoAskMessageTemplate alloc] init] refreshPushPermissions];
+    [[[LPPrePushMessageTemplate alloc] init] refreshPushPermissions];
```
Comment (e7mac) [inline Leanplum-SDK/Classes/LPMessageTemplates.m:104]: If we want. to change the name here, we need to follow it up with changing this on Android + communicating the change of name to the org + documentation

File: Leanplum-SDK/Classes/MessageTemplates/LPAlertMessageTemplate.m [hunk 3]
```diff
@@ -2,46 +2,89 @@
...
+            return YES;
+        }
+        @catch (NSException *exception) {
+            LOG_LP_MESSAGE_EXCEPTION;
+            return NO;
+        }
+    }];
+
+    // Confirm
+    [Leanplum defineAction:LPMT_CONFIRM_NAME
```
Comment (e7mac) [inline Leanplum-SDK/Classes/MessageTemplates/LPAlertMessageTemplate.m:50]: Why are we going back to putting multiple messages in the same file? This makes it hard to get to the right file without previously having context, which is why we started breaking them up into individual files

File: Leanplum-SDK/Classes/LPMessageTemplates.m (hunk 1 above)
Comment (milos1290) [reply Leanplum-SDK/Classes/LPMessageTemplates.m:87]: This just defining of actions, i purposely left it like we had before to better align previous state. The templates are just defining an action that is going to be executed and show appropriate UI. There is no need to have non static methods here as its just adding an action with callback to internal state which will be executed when its triggered.

File: Leanplum-SDK/Classes/LPMessageTemplates.m (hunk 2 above)
Comment (milos1290) [reply Leanplum-SDK/Classes/LPMessageTemplates.m:104]: We can revert it, its just internal change to better describe what is the purpose of the template.

File: Leanplum-SDK/Classes/MessageTemplates/LPAlertMessageTemplate.m (hunk 3 above)
Comment (milos1290) [reply Leanplum-SDK/Classes/MessageTemplates/LPAlertMessageTemplate.m:50]: They are defined in same place as both are using same logic and same ui, just presenting UIAlertViewController, having 4 files to do that doesn't really make sense.

**Comment by milos1290** [comment]
Yeah its kinda hard to follow the changes on github, the codebase looks much more cleaner in the XCode.

**Comment by e7mac** [comment]
Couple of things we need to take care of:
1. We were previously showing messages on top of the visibleViewController which has now switched to root view controller only. This means that if we trigger a message while the rootViewController is presenting another view controller, we won't get that message at all. We should put that back in. I previously had this in the BaseMessageTemplate but since everything is static in this, I don't know the best way to put that back in.

`+ (UIViewController *)visibleViewController
{
    UIViewController *topController = [UIApplication sharedApplication].keyWindow.rootViewController;
    while (topController.presentedViewController) {
        topController = topController.presentedViewController;
    }
    return topController;
}
`
2. We still need to have our snapshot tests working. Each template class populates the data from a context object to build up its view, and the snapshot tests ensure that those work. Previously, the non-static classes would be returning the relevant view or view controller that could be snapshot tested. We should do something similar even with the VCs - the Alert was also previously a VC and we were still able to snapshot test that.

File: Leanplum-SDK/Classes/MessageTemplates/LPMessageTemplateProtocol.h [hunk 4]
```diff
@@ -6,19 +6,17 @@
...
 #import "LPMessageTemplateConstants.h"
+#import "LPApplication+Extensions.h"
 
 NS_ASSUME_NONNULL_BEGIN
 
-@interface LPBaseMessageTemplate : NSObject
+@protocol LPMessageTemplateProtocol
 
-@property  (nonatomic, strong) NSMutableArray *contexts;
+@property (nonatomic, strong) LPActionContext *context;
```
Comment (dejan2k) [inline Leanplum-SDK/Classes/MessageTemplates/LPMessageTemplateProtocol.h:17]: Maybe we can remove the context from the protocol. Because we are not really using this one.
You are using the context define in the view controller or not using the context at all (in cases where we are using alert view controller)

Comment (milos1290) [reply Leanplum-SDK/Classes/MessageTemplates/LPMessageTemplateProtocol.h:17]: yeah, there are some places not used at all, i left it only because of push notifications IAM, as soon as we move it we can remove the context from protocol

Comment (dejan2k) [reply Leanplum-SDK/Classes/MessageTemplates/LPMessageTemplateProtocol.h:17]: So, we can either remove it from the protocol or set it as an optional in the protocol and only use it in templates that we need it

Comment (milos1290) [reply Leanplum-SDK/Classes/MessageTemplates/LPMessageTemplateProtocol.h]: updated
